engine/pkg/dm/message_agent.go

```diff
@@ -0,0 +1,407 @@
+	requestTp
+	responseTp
+)
+
+// message use for asynchronous message and synchronous request/response.
+type message struct {
+	ID      messageID
+	Type    messageType
+	Command string
+	Payload interface{}
+}
+
+// Sender defines an interface that supports send message
+// The caller should register the sender to the messageAgent before call SendMessage/SendRequest.
+type Sender interface {
+	SendMessage(ctx context.Context, topic p2p.Topic, message interface{}, nonblocking bool) error
+}
+
+// messagePair implement a simple synchronous request/response message pattern since the lib currently only support asynchronous message.
+// Caller should persist the request message if needed.
+// Caller should add retry mechanism if needed.
+type messagePair struct {
+	// messageID -> response channel
+	// TODO: limit the MaxPendingMessageCount if needed.
+	pendings sync.Map
+	id       atomic.Uint64
+}
+
+// newMessagePair creates a new MessagePair instance
+func newMessagePair() *messagePair {
+	return &messagePair{}
+}
+
+func (m *messagePair) allocID() messageID {
+	return messageID(m.id.Add(1))
+}
+
+// sendRequest sends a request message and wait for response.
+func (m *messagePair) sendRequest(ctx context.Context, topic p2p.Topic, command string, req interface{}, sender Sender) (interface{}, error) {
+	msg := message{ID: m.allocID(), Type: requestTp, Command: command, Payload: req}
+	respCh := make(chan interface{}, 1)
+	m.pendings.Store(msg.ID, respCh)
+	defer m.pendings.Delete(msg.ID)
+
+	if err := sender.SendMessage(ctx, topic, msg, false /* nonblock */); err != nil {
+		return nil, err
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case resp := <-respCh:
+		return resp, nil
+	}
+}
+
+// sendResponse sends a response with message ID.
+func (m *messagePair) sendResponse(ctx context.Context, topic p2p.Topic, id messageID, command string, resp interface{}, sender Sender) error {
+	msg := message{ID: id, Type: responseTp, Command: command, Payload: resp}
+	return sender.SendMessage(ctx, topic, msg, false /* nonblock */)
+}
+
+// onResponse receives and pairs a response message.
+func (m *messagePair) onResponse(id messageID, resp interface{}) error {
+	respCh, ok := m.pendings.Load(id)
+	if !ok {
+		return errors.Errorf("request %d not found", id)
+	}
+
+	select {
+	case respCh.(chan interface{}) <- resp:
+		return nil
+	default:
+	}
+	return errors.Errorf("duplicated response of request %d, and the last response is not consumed", id)
+}
+
+// MessageAgent defines interface for message communication.
+type MessageAgent interface {
+	Init(ctx context.Context) error
+	Tick(ctx context.Context) error
+	Close(ctx context.Context) error
+	// update the sender to receive message/request.
+	// update the sender before call SendMessage/SendRequest.
+	// update the sender when the sender is close.
+	UpdateSender(senderID string, sender Sender) error
+	SendMessage(ctx context.Context, senderID string, command string, msg interface{}) error
```

Comment:
Now it's clientID. We use taskID instead of workerID as clientID, because for caller, it doesn't care which worker is receiver/sender, it only cares about the task